src/System.Net.HttpListener/src/System/Net/HttpListener.cs

```diff
@@ -163,16 +164,22 @@ internal void AddPrefix(string uriPrefix)
                 {
                     throw new ArgumentException(SR.net_listener_slash, nameof(uriPrefix));
                 }
-                registeredPrefix = uriPrefix[j] == ':' ? string.Copy(uriPrefix) : uriPrefix.Substring(0, j) + (i == 7 ? ":80" : ":443") + uriPrefix.Substring(j);
-                fixed (char* pChar = registeredPrefix)
+                StringBuilder registeredPrefixBuilder = new StringBuilder();
+                if (uriPrefix[j] == ':')
                 {
-                    i = 0;
-                    while (pChar[i] != ':')
-                    {
-                        pChar[i] = (char)CaseInsensitiveAscii.AsciiToLower[(byte)pChar[i]];
-                        i++;
-                    }
+                    registeredPrefixBuilder.Append(uriPrefix);
+                }
+                else
+                {
+                    registeredPrefixBuilder.Append(uriPrefix, 0, j);
+                    registeredPrefixBuilder.Append(i == 7 ? ":80" : ":443");
+                    registeredPrefixBuilder.Append(uriPrefix, j, uriPrefix.Length - j);
+                }
+                for (i = 0; registeredPrefixBuilder[i] != ':'; i++)
+                {
+                    registeredPrefixBuilder[i] = (char)CaseInsensitiveAscii.AsciiToLower[(byte)registeredPrefixBuilder[i]];
```

Comment:
Indexing a StringBuilder can be pretty slow. Could you do this at the beginning, appending the right chars from the get-go?

Reply:
I believe this code path is used only when registering the initial prefixes with the `HttpListener` instance, which is already an enormously slow operation. (The cost of p/invoke and transitioning into kernel mode dominates.) I didn't want to make this code any more complex than it already was given that it's not in a hot code path.

@rmkerr / @davidsh - Can you confirm my understanding here? Should we attempt to optimize this code path?

Reply:
This is  not a hot code path. It is used only right before the actual start of the listener to register the prefixes into http.sys.